src/frolladaptive.c

```diff
@@ -1,32 +1,53 @@
 #include "data.table.h"
 
-/* fast adaptive rolling mean - router
- * algo = 0: fadaptiverollmeanFast
- *   first pass cumsum based solution, second pass uses cumsum to calculate answer
- * algo = 1: fadaptiverollmeanExact
- *   recalculate whole mean for each observation, roundoff correction is adjusted, also support for NaN and Inf
+/* rolling adaptive fun - router for fun and algo
+ * rfun enum rollfun_t routes to rolling function
+ * algo = 0: fast
+ *   adding/removing in/out of sliding window of observations
+ * algo = 1: exact
+ *   recalculate whole fun for each observation, for mean roundoff correction is adjusted, also support for NaN and Inf
  */
-void fadaptiverollmean(unsigned int algo, double *x, uint64_t nx, ans_t *ans, int *k, double fill, bool narm, int hasna, bool verbose) {
+void frolladaptivefun(rollfun_t rfun, unsigned int algo, double *x, uint64_t nx, ans_t *ans, int *k, double fill, bool narm, int hasna, bool verbose) {
   double tic = 0;
   if (verbose)
     tic = omp_get_wtime();
-  if (algo==0) {
-    fadaptiverollmeanFast(x, nx, ans, k, fill, narm, hasna, verbose);
-  } else if (algo==1) {
-    fadaptiverollmeanExact(x, nx, ans, k, fill, narm, hasna, verbose);
+  switch (rfun) {
+  case MEAN :
+    if (algo==0) {
+      frolladaptivemeanFast(x, nx, ans, k, fill, narm, hasna, verbose);
+    } else if (algo==1) {
+      frolladaptivemeanExact(x, nx, ans, k, fill, narm, hasna, verbose);
+    }
+    break;
+  case SUM :
+    if (algo==0) {
+      frolladaptivesumFast(x, nx, ans, k, fill, narm, hasna, verbose);
+    } else if (algo==1) {
+      frolladaptivesumExact(x, nx, ans, k, fill, narm, hasna, verbose);
+    }
+    break;
+  case MAX :
+    if (algo==0 && verbose) {
+      //frolladaptivemaxFast(x, nx, ans, k, fill, narm, hasna, verbose); // frolladaptivemaxFast does not exists as of now
+      snprintf(end(ans->message[0]), 500, _("%s: algo %u not implemented, fall back to %u\n"), __func__, algo, (unsigned int) 1);
+    }
+    frolladaptivemaxExact(x, nx, ans, k, fill, narm, hasna, verbose);
+    break;
+  default:
+    error(_("Internal error: Unknown rfun value in froll: %d"), rfun); // #nocov
   }
   if (verbose)
-    snprintf(end(ans->message[0]), 500, _("%s: processing algo %u took %.3fs\n"), __func__, algo, omp_get_wtime()-tic);
-  // implicit n_message limit discussed here: https://github.com/Rdatatable/data.table/issues/3423#issuecomment-487722586
+    snprintf(end(ans->message[0]), 500, _("%s: processing fun %d algo %u took %.3fs\n"), __func__, rfun, algo, omp_get_wtime()-tic);
```

Comment:
`frolladaptivefun: algo 0 not implemented, fall back to 1` might not be the most informative message

Reply:
Completely agree that it is not user friendly. Reason why it was added so that we could test if expected algo is being used. Rolling functions do not expose `verbose` arg directly, it has to be set via options, therefore full user friendliness is maybe not the biggest priority at the moment.
We will also have to keep in mind buffer limit of 500 characters set for `snprintf`. Nevertheless I will note down "improving friendliness of verbose messages" as a post change, together with `test(options=verbose)` as both will require many tests to be updated.